### ISSUE TITLE
OpenAICompletion doesn't work for JSON parsing? 

### ISSUE THREAD

**Issue by jerryjliu** [outlines-dev/outlines#227]
I tried plugging in OpenAICompletion instead of the default `transformers('gpt-2')` model shown in the readme, for the JSON parsing example. But i'm running into the following error:

Code:
```
import outlines.models as models
import outlines.text.generate as generate

# model = models.transformers("gpt2-medium")
model = models.OpenAICompletion("gpt-3.5-turbo", max_tokens=256, temperature=0.1)

sequence = generate.json(model, Character)("Give me a character description")
```

Error:

<img width="1169" alt="Screenshot 2023-08-15 at 8 58 41 AM" src="https://github.com/normal-computing/outlines/assets/4858925/a17ff69a-444f-4eae-8d50-df36456d5f2b">

**Comment by mondaychen** [comment]
I'll let Rémi add details, but to my knowledge controlled generation only works OK on limited case (e.g. `type="int"`) due to restrictions on OpenAI’s API.

**Comment by rlouf** [comment]
Long story short: you can't.

Now the long story. To be able to constrain the generation you need to be able to mask the logits. However, the OpenAI API only allows you to mask 300 tokens for each requests, while GPT4's vocabulary is ~100k tokens long.
Having said that, our method allows to determine in advance how many tokens you will need to mask, so there may be cases where we can rigorously do constrained generation with the OpenAI API, at least warn the users that the results won't be correct when that's not the case. 

If you would still find it useful, let me know! See #157 as well.

**Comment by trevor-tocchet-picket** [comment]
So no chat completion models can be used with JSON generation? @rlouf

**Comment by rlouf** [comment]
OSS ones definitely can be used

**Comment by trevor-tocchet-picket** [comment]
But until OpenAI API allows a great mask for each request then we cannot use models like GPT-4 etc.

**Comment by shabie** [comment]
So to my understanding: Since we can only mask 300 tokens (by giving them a value of -100), we can't use OpenAI's API as is since the vocabulary is too big and many tokens that are invalid will still get a non-zero probability.

One way to addressthis: Due to the fact that we have regex, we can say how many tokens _need_ to be masked out (i.e. everything that doesn't match) and if that exceeds 300, we can warn the user of unreliable results.

Did I get this right?

**Comment by rlouf** [comment]
> Did I get this right?

Absolutely!
